internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,190 @@
+	itLock     sync.Mutex
+	notifyChan chan bool
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	return &logGatherer{
+		target:     target,
+		tickPeriod: 1 * time.Second,
+		cancel:     make(chan struct{}),
+		client:     client,
+		notifyChan: make(chan bool, 1),
+	}, nil
+}
+
+func (g *logGatherer) addIterator(serviceName string, iterator servicelog.Iterator, parser *servicelog.Parser) {
+	g.itLock.Lock()
+	defer g.itLock.Unlock()
+
+	iterator.Notify(g.notifyChan)
+	g.iterators[serviceName] = iterator
+
+	if parser == nil {
+		parser = servicelog.NewParser(iterator, parserSize)
+	}
+	g.parsers[serviceName] = parser
+}
+
+// Clear all iterators/parsers and return the old ones.
+func (g *logGatherer) clearIterators() (map[string]servicelog.Iterator, map[string]*servicelog.Parser) {
+	g.itLock.Lock()
+	defer g.itLock.Unlock()
+
+	oldIterators := make(map[string]servicelog.Iterator, len(g.iterators))
+	for svc, it := range g.iterators {
+		oldIterators[svc] = it
+	}
+	g.iterators = map[string]servicelog.Iterator{}
+
+	oldParsers := make(map[string]*servicelog.Parser, len(g.parsers))
+	for svc, p := range g.parsers {
+		oldParsers[svc] = p
+	}
+	g.parsers = map[string]*servicelog.Parser{}
+
+	return oldIterators, oldParsers
+}
+
+func (g *logGatherer) loop() {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush()
+			if err != nil {
+				logger.Noticef("Cannot send logs to target %q: %v", g.target.Name, err)
+			}
+
+		case <-g.cancel:
+			// Gatherer has been stopped - flush any remaining logs
+			err := g.client.Flush()
+			if err != nil {
+				logger.Noticef("Cannot send logs to target %q: %v", g.target.Name, err)
+			}
+			return
+
+		case <-g.notifyChan:
+			// (at least) one of the iterators has data to read
+			g.pullLogs()
+		}
+	}
+}
+
+// stop closes the cancel channel, thereby terminating the main loop.
+func (g *logGatherer) stop() {
+	close(g.cancel)
+}
+
+func (g *logGatherer) pullLogs() {
+	g.itLock.Lock()
+	defer g.itLock.Unlock()
+
+	moreData := false
+	for svc, it := range g.iterators {
+		if !it.Next(nil) {
+			continue
+		}
+		parser := g.parsers[svc]
+		if !parser.Next() {
+			continue
+		}
+		if err := parser.Err(); err != nil {
+			// TODO: handle this properly ?
+			logger.Noticef("Cannot read logs from service %q: %v", svc, err)
+			continue
+		}
+
+		entry := parser.Entry()
+		err := g.client.Write(entry)
+		if err != nil {
+			logger.Noticef("Cannot write logs to target %q: %v", g.target.Name, err)
+		}
+
+		if it.Next(nil) {
+			moreData = true
+		}
+	}
+
+	// If there is more data to read - put a token back in the notifyChan, so
+	// that we will continue to pull logs on the next loop iteration.
+	if moreData {
+		select {
+		case g.notifyChan <- true:
```

Comment:
Again, this looks a bit brittle. If we process all the data currently in the iterator, we shoudn't have to feed our own watcher because we can only do something once we have more data, at which point the iterator itself will notify us that there's more data. If we hook up that pipeline between both sides in a nice way, this middle layer should get pretty "boring", in the good sense.